[Elastic Agent] Add doc on how Fleet Server is bootstrapped. (#29563)
elastic/beats: x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc

[[fleet-server-bootstrap]]
== Fleet Server Bootstrap

Elastic Agent with Fleet Server has a bootstrap process that it uses to get
Fleet Server up and running under Elastic Agent.

Upon `install` or `enroll` with Elastic Agent will bootstrap a Fleet Server
when the `--fleet-server-es` command-line option is provided. In this mode
Elastic Agent will only communicate with its local Fleet Server and expose

> the Fleet Server over the `:8220` port.

Reviewer: Based on this: https://github.com/elastic/fleet-server/pull/967/files
Both 8220 and 8221 will be used. 8221 for exclusive communication with the local Elastic Agent running.

Author: I updated to documentation to use :8221 and added description on the difference between :8220 and :8221.

[[fleet-server-bootstrap]]
== Fleet Server Bootstrap

Elastic Agent with Fleet Server has a bootstrap process that it uses to get
Fleet Server up and running under Elastic Agent.

> Upon `install` or `enroll` with Elastic Agent will bootstrap a Fleet Server
> when the `--fleet-server-es` command-line option is provided. In this mode

Reviewer 2: ```suggestion
Elastic Agent will bootstrap a Fleet Server when the `--fleet-server-es`
command-line option is provided to an `install` or `enroll` command
```

i think it's a bit easier to read

Elastic Agent will only communicate with its local Fleet Server and expose
the Fleet Server over the `:8221` port.

> `:8221` is only bound to the localhost of the machine/container and cannot
> be accessed remotely. The `:8220` port is also bound but is bound on `0.0.0.0`
> to allow remote connections from external Elastic Agents that wish to enroll
> and communication. The `:8221` port is reserved for communication between the
> Elastic Agent and the Fleet Server on the host. This ensures that the local
> Elastic Agent has priority in check-ins with the Fleet Server.

Reviewer 2: ```suggestion
The `:8221` port is reserved for communication between the Elastic Agent and the
Fleet Server on the host. It is bound to the localhost of the machine/container
and cannot be accessed remotely. This ensures that the local Elastic Agent has
priority in check-ins with the Fleet Server. The `:8220` port is bound on
`0.0.0.0` to allow remote connections from external Elastic Agents that wish to
enroll and communication.
```
re ordered to make it clear what the differences between ports are

[float]
[[fleet-server-operating-modes]]
=== Operating Modes

> Elastic Agent can bootstrap the Fleet Server into three different modes. In either
> mode the Fleet Server operates the same, but how it exposes itself over the `:8220`
> port is different.

Reviewer 2: ```suggestion
Elastic Agent can bootstrap the Fleet Server into three different modes. The mode
determines how Fleet Server exposes itself over the `:8220` port, but does not change
any other behaviour.
```

==== Self-signed Certificate

With the standard `--fleet-server-es` and `--fleet-server-service-token` options the

> Elastic Agent will generate a self-generated CA and certificate for communication with
> the Fleet Server that it starts under the Elastic Agent. These certificates are generated

Reviewer 2: ```suggestion
Elastic Agent will generate a CA and certificate for communication with
the Fleet Server that it starts. These certificates are generated
```

by Elastic Agent and passed to the Fleet Server, with Elastic Agent using the host's
hostname in the communication URL for valid TLS verification.

==== HTTP Only

Using the `--insecure` and `--fleet-server-insecure-http` will bootstrap the Fleet Server
without any certificates, it will be bound to `localhost:8220` and Elastic Agent will
communicate in clear-text.

==== Custom Certificates (aka. Production)

> When deploying Elastic Agent in a production environment using enterprise generated

Reviewer 3: s/enterprise generated/enterprise-generated/

certificates will ensure that Elastic Agent running locally and remote Elastic Agent

> will be able to connect over a verified TLS based connection. Passing in the certificates
> use `--fleet-server-cert`, `--fleet-server-cert-ca`, and `--certificate-authorities`.

Reviewer 2: ```suggestion
will be able to connect over a verified TLS based connection. Certificates are specified
with `--fleet-server-cert`, `--fleet-server-cert-ca`, and `--certificate-authorities`.
```

[float]
[[fleet-server-bootstrap-process]]
=== How Does It Bootstrap

Bootstrapping happens inside of the `enroll` command. Even when the `install` command
or the `container` command (used by Docker container) is used they both still call

> the `enroll` command to perform the bootstrapping process.

Reviewer 2: ```suggestion
Bootstrapping is ran during the `enroll` command. The `install` command
or `container` command (used by Docker container) will call the `enroll`
command to perform the bootstrapping process.
